src/internal/request-batcher.ts

```diff
@@ -0,0 +1,126 @@
+import { ConnectionInfo, FetchJsonResponse } from '@ethersproject/web';
+
+import { DEFAULT_MAX_REQUEST_BATCH_SIZE } from '../util/const';
+import { JsonRpcRequest } from './internal-types';
+
+/**
+ * Internal class to enqueue requests and automatically send/process batches.
+ *
+ * The underlying batching mechanism is loosely based on ethers.js's
+ * `JsonRpcBatchProvider`.
+ *
+ * @internal
+ */
+export class RequestBatcher {
+  /** Timeout timer that periodically sends the pending batch. */
+  private pendingBatchTimer: NodeJS.Timer | undefined;
+
+  /**
+   * Array of enqueued requests along with the constructed promise handlers for
+   * each request.
+   */
+  private pendingBatch: Array<BatchRequest> | undefined;
+
+  constructor(
+    private readonly sendBatchFn: SendBatchFn,
+    private readonly connection: ConnectionInfo,
+    private readonly maxBatchSize = DEFAULT_MAX_REQUEST_BATCH_SIZE
+  ) {}
+
+  /**
+   * Enqueues the provided request. The batch is immediately sent if the maximum
+   * batch size is reached. Otherwise, the request is enqueued onto a batch that
+   * is sent after 10ms.
+   *
+   * Returns a promise that resolves with the result of the request.
+   */
+  async enqueueRequest(request: JsonRpcRequest): Promise<any> {
+    if (this.pendingBatch === undefined) {
+      this.pendingBatch = [];
+    }
+
+    const inflightRequest: BatchRequest = {
+      request,
+      resolve: undefined,
+      reject: undefined
+    };
+
+    const promise = new Promise((resolve, reject) => {
+      inflightRequest.resolve = resolve;
+      inflightRequest.reject = reject;
+    });
+
+    this.pendingBatch.push(inflightRequest);
+
+    if (this.pendingBatch.length === this.maxBatchSize) {
+      // Send batch immediately if we are at the maximum batch size.
+      void this.sendBatchRequest();
+    } else if (!this.pendingBatchTimer) {
+      // Schedule batch for next event loop + short duration
+      this.pendingBatchTimer = setTimeout(() => this.sendBatchRequest(), 10);
+    }
+
+    return promise;
+  }
+
+  /**
+   * Sends the currently queued batches and resets the batch and timer. Processes
+   * the batched response results back to the original promises.
+   */
+  private async sendBatchRequest(): Promise<void> {
+    // This if-statement handles the case where the maximum batch size triggers
+    // the batch send, so the scheduled timeout send is called with an empty
+    // batch (or vice-versa).
+    if (this.pendingBatch === undefined) {
+      return;
+    }
+
+    // Get the current batch and clear it, so new requests
+    // go into the next batch
+    const batch = this.pendingBatch!;
+    this.pendingBatch = undefined;
+    this.pendingBatchTimer = undefined;
```

Comment:
You might want to `clearTimeout` on the timer here. That way if the batch send is triggered by the size limit, then it won't try to send a partial batch until a full 10 ms have passed.